File: shinyJSME-master/inst/htmlwidgets/jsme.js
(function() {
  var readMolecule = function(el, structure) {
    $('#'+el).data('jsme').readMolecule(structure);
  };

  var readMolFile = function(el, structure) {
    $('#'+el).data('jsme').readMolFile(structure);
  };

  var reset = function(el) {
    $('#'+el).data('jsme').reset();
  };

  var smiles = function(el, inputEl) {
    Shiny.onInputChange(inputEl, $('#'+el).data('jsme').smiles());
  };

  var molFile = function(el, inputEl) {
    Shiny.onInputChange(inputEl, $('#'+el).data('jsme').molFile());
  };

  var jmeFile = function(el, inputEl) {
    Shiny.onInputChange(inputEl, $('#'+el).data('jsme').jmeFile());
  };

  Shiny.addCustomMessageHandler('useJME',
    function(message) {
      readMolecule(message.el, message.jmeFile);
    }
  );

  Shiny.addCustomMessageHandler('useMOL',
    function(message) {
      readMolFile(message.el, message.molFile);
    }
  );

  Shiny.addCustomMessageHandler('resetEditor',
    function(message) {
      reset(message.el);
    }
  );

  Shiny.addCustomMessageHandler('smiles',
    function(message) {
      smiles(message.el, message.inputEl);
    }
  );

  Shiny.addCustomMessageHandler('molFile',
    function(message) {
      molFile(message.el, message.inputEl);
    }
  );

  Shiny.addCustomMessageHandler('jmeFile',
    function(message) {
      jmeFile(message.el, message.inputEl);
    }
  );
})();



HTMLWidgets.widget({

  name: 'jsme',

  type: 'output',

  factory: function(el, width, height) {



    if (typeof window.jsmeArray === 'undefined') window.jsmeArray = [];
    window.jsmeArray.push({el: el.id, width: width, height: height});

    jsmeOnLoad = function() {
      if (typeof window.jsmeArray !== 'undefined') {
        window.jsmeArray.forEach(
          function(v) {
            if (typeof $('#'+v.el).data('jsme') === 'undefined') {
              $('#'+v.el).data('jsme', new JSApplet.JSME(v.el, '100%', '100%'));
              $('#'+v.el).data('jsme').setSize(v.width, v.height);
            }
          }
        );
      }
	  };


    return {

      renderValue: function() {

        if ((typeof $('#'+el.id).data('jsme') === 'undefined') && (typeof JSApplet !== 'undefined')) {
          $('#'+el.id).data('jsme', new JSApplet.JSME(el.id, '100%', '100%')); // jsmeOnLoad();
        }

      },

      resize: function(width, height) {

        $('#'+el.id).data('jsme').setSize(width, height);

      }

    };

  }

});

File: shinyJSME-master/R/jsme.R
#' <Add Title>
#'
#' <Add Description>
#'
#' @import htmlwidgets
#'
#' @export
jsme <- function(structure='', width = '650px', height = '650px') {

  # forward options using x
  x = list(
    structure = structure
  )

  # create widget
  htmlwidgets::createWidget(
    name = 'jsme',
    x = x,
    width = width,
    height = height,
    package = 'jsme'
  )
}

smiles <- function(session, element, inputEl) {
  session$sendCustomMessage(type = 'smiles',
                            message = list(el=element, inputEl = inputEl))
}

molFile <- function(session, element, inputEl) {
  session$sendCustomMessage(type = 'molFile',
                            message = list(el=element, inputEl = inputEl))
}

jmeFile <- function(session, element, inputEl) {
  session$sendCustomMessage(type = 'jmeFile',
                            message = list(el=element, inputEl = inputEl))
}

useMOL <- function(session, element, molFile) {
  session$sendCustomMessage(type = 'useMOL',
                            message = list(el=element, molFile = molFile))
}

useJME <- function(session, element, jmeFile) {
  session$sendCustomMessage(type = 'useJME',
                            message = list(el=element, jmeFile = jmeFile))
}

resetEditor <- function(session, element) {
  session$sendCustomMessage(type = 'resetEditor',
                            message = list(el=element))
}

#' Shiny bindings for jsme
#'
#' Output and render functions for using jsme within Shiny
#' applications and interactive Rmd documents.
#'
#' @param outputId output variable to read from
#' @param width,height Must be a valid CSS unit (like \code{'100\%'},
#'   \code{'400px'}, \code{'auto'}) or a number, which will be coerced to a
#'   string and have \code{'px'} appended.
#' @param expr An expression that generates a jsme
#' @param env The environment in which to evaluate \code{expr}.
#' @param quoted Is \code{expr} a quoted expression (with \code{quote()})? This
#'   is useful if you want to save an expression in a variable.
#'
#' @name jsme-shiny
#'
#' @export
jsmeOutput <- function(outputId, width = '100px', height = '100px'){
  htmlwidgets::shinyWidgetOutput(outputId, 'jsme', width, height, package = 'jsme')
}

#' @rdname jsme-shiny
#' @export
renderJsme <- function(expr, env = parent.frame(), quoted = FALSE) {
  if (!quoted) { expr <- substitute(expr) } # force quoted
  htmlwidgets::shinyRenderWidget(expr, jsmeOutput, env, quoted = TRUE)
}
